Exception-Handling.md

```diff
@@ -0,0 +1,252 @@
+
+The syntax will look like this:
+
+```ruby
+begin
+  # some code that will create an exception.
+  
+rescue <some kind of exception>
+  # Code to recover from that exception
+
+rescue <some other kind of exception>
+  # Code to recover from another type of exception
+
+rescue
+  # other exceptions handled here
+
+ensure
+  #  This code runs no matter what.  
+
+end 
+```
+#### One Ruby Example - CSVReader
+
+In Ruby we could write code to handle problems, like in our CSV code to deal with problems, like being unable to open the CSV file.
+
+```ruby
+content = []
+begin
+  CSV.foreach("file.csv") do |row|
+
+    content << row
+  end
+
+  rescue SystemCallError
+    puts "I couldn't open the file."
+
+end
+```
+
+#### Exceptions Continue Up The Stack
+
+When an exception occurs, if it is not rescued in the immediate block or method it proceeds up and out of the method until the application crashes with an error message.
+
+```ruby
+def quotient_and_remainer dividend, divisor
+  answer = [dividend / divisor]
+  answer << quotient % divisor
+  answer
+end
+
+begin
+  puts quotient_and_remainer 10, 0
+rescue ZeroDivisionError
+  puts "Attempted to divide by zero"
+end
+```
+The method `divide` here will generate an exception when it tries to divide 10 by zero.  The exception is not "caught" in the method so ruby leaves the method and falls back to the caller of the method, which does have a rescue block to catch the `ZeroDivisionError`.  
+
+If the rescue block was missing the resulting error message would be:
+
+```bash
+method_exception_example.rb:3:in `/': divided by 0 (ZeroDivisionError)
+	from method_exception_example.rb:3:in `divide'
+	from method_exception_example.rb:10:in `<main>'
+```
+
+This is the result of the call stack.  The Error first occurs at line 3 in the 'divide' method (see above).  When the block is not "rescued," Ruby leaves the divide method and falls back to the caller of the method, and so on until it reaches the "main" program and exits.  Each line of the result, called a "backtrace," shows the path of execution in reverse order.  
+
+So in the following example:
+
+```ruby
+def quotient_and_remainer dividend, divisor
+  answer = [dividend / divisor]
+  answer << dividend % divisor
+  answer
+end
+
+def mystery(num)
+  answer = []
+
+  (0...9).each do |i|
+    answer << quotient_and_remainer(num, i)
+  end
+  answer
+end
+
+begin
+  puts "The Remainders are #{mystery(100)}"
+end
+```
+
+1.  Ruby Starts by calling the `mystery` method which starts a loop with values from 0 to 9 (inclusive).  
+2.  It then calls the `quotient_and__remainer` method with 100 & 0 as arguments.  
+3.  Inside the quotient\_and\_remainder method it has a divide\_by\_zero error.
```

Comment:
For consistency we should use `quotient_and_remainder` here, and call the error `ZeroDivisionError`.